File: go/v2/v2_wire_test.go
package openwopclient

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

type capturedRequest struct {
	Method string
	Path   string
	Query  string
	Header http.Header
}

// newWireServer records every request and answers with the given status +
// body (or per-path bodies when routes is non-nil).
func newWireServer(t *testing.T, status int, body string, routes map[string]string) (*httptest.Server, *[]capturedRequest) {
	t.Helper()
	captured := &[]capturedRequest{}
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		*captured = append(*captured, capturedRequest{Method: r.Method, Path: r.URL.Path, Query: r.URL.RawQuery, Header: r.Header.Clone()})
		out := body
		code := status
		if routes != nil {
			if b, ok := routes[r.URL.Path]; ok {
				out = b
			}
		}
		if strings.HasPrefix(out, "404:") {
			code = http.StatusNotFound
			out = strings.TrimPrefix(out, "404:")
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(code)
		_, _ = w.Write([]byte(out))
	}))
	t.Cleanup(srv.Close)
	return srv, captured
}

func TestProtocolVersionHeaderOnEveryRequest(t *testing.T) {
	if SDKProtocolMajor != 2 || ProtocolVersionHeader(2) != "2.0" || ProtocolVersionHeader(3) != "3.0" {
		t.Fatal("major rendering mismatch")
	}
	srv, captured := newWireServer(t, 200, `{}`, map[string]string{
		"/.well-known/openwop": `{"protocolVersions":["2.0"],"preferredVersion":"2.0"}`,
		"/runs":                `{"runId":"t/r","status":"pending","eventsUrl":"/runs/t%2Fr/events"}`,
	})
	client, err := NewClient(srv.URL, "k")
	if err != nil {
		t.Fatal(err)
	}
	if client.ProtocolVersion() != "2.0" {
		t.Fatalf("default ProtocolVersion: %q", client.ProtocolVersion())
	}
	ctx := context.Background()
	if _, err := client.GetCapabilities(ctx); err != nil {
		t.Fatal(err)
	}
	if _, err := client.GetWorkflow(ctx, "wf"); err != nil {
		t.Fatal(err)
	}
	if _, err := client.GetOpenAPI(ctx); err != nil {
		t.Fatal(err)
	}
	if _, err := client.CreateRun(ctx, CreateRunRequest{WorkflowID: "wf"}, MutationOptions{IdempotencyKey: "idem-1", Dedup: true}); err != nil {
		t.Fatal(err)
	}
	if len(*captured) != 4 {
		t.Fatalf("expected 4 requests, got %d", len(*captured))
	}
	for _, req := range *captured {
		if req.Header.Get("OpenWOP-Version") != "2.0" {
			t.Errorf("%s %s: OpenWOP-Version=%q", req.Method, req.Path, req.Header.Get("OpenWOP-Version"))
		}
		if strings.Contains(req.Path, "/v1") {
			t.Errorf("%s %s: versioned path", req.Method, req.Path)
		}
	}
	reqs := *captured
	if reqs[0].Header.Get("Authorization") != "" || reqs[2].Header.Get("Authorization") != "" {
		t.Error("discovery + openapi are unauthenticated")
	}
	if reqs[1].Header.Get("Authorization") != "Bearer k" {
		t.Error("workflow read must carry the bearer key")
	}
	create := reqs[3]
	if create.Path != "/runs" || create.Header.Get("OpenWOP-Dedup") != "enforce" || create.Header.Get("Idempotency-Key") != "idem-1" || create.Header.Get("X-Dedup") != "" {
		t.Errorf("create headers/path mismatch: %+v", create)
	}

	client.Major = 3
	if _, err := client.GetWorkflow(ctx, "wf"); err != nil {
		t.Fatal(err)
	}
	if (*captured)[4].Header.Get("OpenWOP-Version") != "3.0" {
		t.Error("Major must drive the header")
	}
}

func TestProtocolVersionUnsupportedSurfacesTypedCode(t *testing.T) {
	srv, _ := newWireServer(t, http.StatusNotAcceptable, `{"error":"protocol_version_unsupported","message":"unlisted","details":{"protocolVersions":["1.12"]}}`, nil)
	client, _ := NewClient(srv.URL, "k")
	_, err := client.GetRun(context.Background(), "t/r1")
	var werr *WopError
	if !errors.As(err, &werr) || werr.Status != 406 || werr.Envelope == nil || werr.Envelope.Error != "protocol_version_unsupported" {
		t.Fatalf("expected a typed 406, got %v", err)
	}
	if !IsErrorCode(werr.Envelope.Error) {
		t.Fatal("protocol_version_unsupported must be a registered code")
	}
}

func TestUnversionedPathsAcrossNamespaces(t *testing.T) {
	// r.URL.Path is the decoded path, so the routes use "/" not "%2F".
	srv, captured := newWireServer(t, 200, `{}`, map[string]string{
		"/runs/t/r1/compensation": `{"runId":"t/r1","status":"none","plan":[],"attempts":[]}`,
		"/runs/t/r1/effects":      `404:{"error":"not_found","message":"no"}`,
		"/host/effect-seams":      `{"manifestVersion":"1","host":{"name":"h","build":{"kind":"commit","id":"abc"}},"seams":[{"seam":"http.fetch","kind":"http","guarded":true,"guardedBy":"interceptor"}]}`,
		"/runs/t/r1/events/poll":  `{"runId":"t/r1","events":[],"lastSequence":-1,"status":"running","isTerminal":false}`,
	})
	client, _ := NewClient(srv.URL, "k")
	ctx := context.Background()
	comp, err := client.GetRunCompensation(ctx, "t/r1")
	if err != nil || comp == nil || comp.Status != "none" {
		t.Fatalf("compensation: %v %+v", err, comp)
	}
	eff, err := client.GetRunEffects(ctx, "t/r1")
	if err != nil || eff != nil {
		t.Fatalf("effects 404 must be (nil, nil): %v %+v", err, eff)
	}
	seams, err := client.GetEffectSeamManifest(ctx)
	if err != nil || len(seams.Seams) != 1 || seams.Seams[0].Seam != "http.fetch" {
		t.Fatalf("seams: %v %+v", err, seams)
	}
	after := 7
	timeout := 5
	page, err := client.PollRunEvents(ctx, "t/r1", PollRunEventsOptions{AfterSequence: &after, TimeoutSeconds: &timeout})
	if err != nil || page.LastSequence != -1 || page.IsTerminal {
		t.Fatalf("poll: %v %+v", err, page)
	}
	if _, err := client.GetAgentOrgChart(ctx); err != nil {
		t.Fatal(err)
	}
	if err := client.DeletePromptTemplate(ctx, "p1"); err != nil {
		t.Fatal(err)
	}
	if err := client.UnregisterWebhook(ctx, "wh1"); err != nil {
		t.Fatal(err)
	}

	reqs := *captured
	want := []string{"/runs/t/r1/compensation", "/runs/t/r1/effects", "/host/effect-seams", "/runs/t/r1/events/poll", "/agents/org-chart", "/prompts/p1", "/webhooks/wh1"}
	if len(reqs) != len(want) {
		t.Fatalf("expected %d requests, got %d", len(want), len(reqs))
	}
	for i, w := range want {
		if reqs[i].Path != w {
			t.Errorf("request %d: path %q want %q", i, reqs[i].Path, w)
		}
	}
	poll := reqs[3].Query
	if !strings.Contains(poll, "afterSequence=7") || !strings.Contains(poll, "timeout=5") || strings.Contains(poll, "lastSequence") {
		t.Errorf("poll query: %q", poll)
	}
}

func TestSSEChannelsCarryTheVersionHeader(t *testing.T) {
	var seen []capturedRequest
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		seen = append(seen, capturedRequest{Method: r.Method, Path: r.URL.Path, Query: r.URL.RawQuery, Header: r.Header.Clone()})
		w.Header().Set("Content-Type", "text/event-stream")
		w.WriteHeader(200)
		if r.URL.Path == "/host/events" {
			_, _ = w.Write([]byte(": keep-alive\n\nevent: heartbeat.evaluated\ndata: {\"type\":\"heartbeat.evaluated\",\"payload\":{\"heartbeatId\":\"h1\",\"status\":\"ok\",\"changed\":false}}\n\n"))
			return
		}
		_, _ = w.Write([]byte("id: 1\nevent: batch\ndata: [{\"eventId\":\"e1\",\"runId\":\"t/r1\",\"type\":\"run.started\",\"payload\":{},\"timestamp\":\"t\",\"sequence\":0,\"schemaVersion\":1},{\"eventId\":\"e2\",\"runId\":\"t/r1\",\"type\":\"run.completed\",\"payload\":{},\"timestamp\":\"t\",\"sequence\":1,\"schemaVersion\":1}]\n\n"))
	}))
	defer srv.Close()
	client, _ := NewClient(srv.URL, "k")
	ctx := context.Background()

	events, cleanup, err := client.StreamEvents(ctx, "t/r1", StreamEventsOptions{StreamModes: []StreamMode{StreamModeUpdates, StreamModeMessages}, LastEventID: "0"})
	if err != nil {
		t.Fatal(err)
	}
	var types []string
	for ev := range events {
		types = append(types, ev.Type)
	}
	cleanup()
	if strings.Join(types, ",") != "run.started,run.completed" {
		t.Fatalf("batch frames must flatten: %v", types)
	}

	host, cleanup2, err := client.StreamHostEvents(ctx, HostEventsOptions{})
	if err != nil {
		t.Fatal(err)
	}
	var hostTypes []string
	for ev := range host {
		hostTypes = append(hostTypes, ev.Type)
	}
	cleanup2()
	if strings.Join(hostTypes, ",") != "heartbeat.evaluated" {
		t.Fatalf("host events: %v", hostTypes)
	}

	if len(seen) != 2 || seen[0].Path != "/runs/t/r1/events" || seen[1].Path != "/host/events" {
		t.Fatalf("unexpected subscribe paths: %+v", seen)
	}
	for _, req := range seen {
		if req.Header.Get("OpenWOP-Version") != "2.0" {
			t.Errorf("%s: OpenWOP-Version=%q", req.Path, req.Header.Get("OpenWOP-Version"))
		}
	}
	if !strings.Contains(seen[0].Query, "streamMode=updates%2Cmessages") || seen[0].Header.Get("Last-Event-ID") != "0" {
		t.Errorf("runEvents subscribe: query=%q Last-Event-ID=%q", seen[0].Query, seen[0].Header.Get("Last-Event-ID"))
	}
}

File: sdk/typescript-v2/src/__tests__/v2-wire.test.ts
/**
 * The v2 wire contract this package exists for (RFC 0172 §A, RFC 0171 §C.1,
 * events.md §Poll): bare-origin unversioned paths, `OpenWOP-Version` on every
 * request, the `OpenWOP-*` header family, `afterSequence`, and the generated
 * error-code registry.
 */

import { describe, it, expect } from 'vitest';
import { OpenwopClient, SDK_PROTOCOL_MAJOR, protocolVersionHeader } from '../client.js';
import { streamEvents, streamHostEvents } from '../sse.js';
import {
  ERROR_CODES,
  ERROR_CODE_HTTP_STATUS,
  RETRIABLE_ERROR_CODES,
  CAPABILITY_FAMILY_KEYS,
  CAPABILITY_METADATA_KEYS,
} from '../generated.js';
import { isErrorCode, isRetriableErrorCode, isVendorErrorCode, HTTP_ERROR_CODES } from '../run-helpers.js';
import { WopError } from '../types.js';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO = resolve(__dirname, '..', '..', '..', '..');

interface Captured { url: string; method: string; headers: Headers }

function mockClient(
  responder: (req: Captured) => { status: number; body?: unknown; headers?: Record<string, string> } = () => ({ status: 200, body: {} }),
  opts: { major?: number } = {},
): { client: OpenwopClient; captured: Captured[] } {
  const captured: Captured[] = [];
  const mockFetch: typeof fetch = async (input, init) => {
    const url = typeof input === 'string' ? input : input instanceof URL ? input.toString() : input.url;
    const req = { url, method: init?.method ?? 'GET', headers: new Headers(init?.headers ?? {}) };
    captured.push(req);
    const r = responder(req);
    const headers = new Headers(r.headers ?? {});
    if (r.body !== undefined && !headers.has('content-type')) headers.set('content-type', 'application/json');
    return new Response(r.body !== undefined ? JSON.stringify(r.body) : null, { status: r.status, headers });
  };
  const client = new OpenwopClient({ baseUrl: 'https://host.example/', apiKey: 'k', fetch: mockFetch, ...opts });
  return { client, captured };
}

describe('OpenWOP-Version negotiation (RFC 0172 §A.3)', () => {
  it('defaults to major 2 and renders <major>.0', () => {
    expect(SDK_PROTOCOL_MAJOR).toBe(2);
    expect(protocolVersionHeader(2)).toBe('2.0');
    expect(protocolVersionHeader(3)).toBe('3.0');
    expect(() => protocolVersionHeader(1.5)).toThrow(TypeError);
    expect(mockClient().client.protocolVersion).toBe('2.0');
    expect(mockClient(undefined, { major: 3 }).client.protocolVersion).toBe('3.0');
  });

  it('is sent on every request — authenticated, unauthenticated, mutation, read', async () => {
    const { client, captured } = mockClient(() => ({ status: 200, body: { protocolVersions: ['2.0'], preferredVersion: '2.0' } }));
    await client.discovery.capabilities();
    await client.runs.get('t/r1');
    await client.runs.create({ workflowId: 'wf' }, { idempotencyKey: 'idem-1', dedup: 'enforce' });
    await client.interrupts.inspectByToken('tok');
    await client.host.effectSeams();
    expect(captured).toHaveLength(5);
    for (const req of captured) {
      expect(req.headers.get('OpenWOP-Version')).toBe('2.0');
      expect(req.url.includes('/v1')).toBe(false);
    }
    expect(captured[0]!.headers.has('Authorization')).toBe(false);
    expect(captured[3]!.headers.has('Authorization')).toBe(false);
    expect(captured[1]!.headers.get('Authorization')).toBe('Bearer k');
  });

  it('a 406 protocol_version_unsupported surfaces as a WopError with the typed code', async () => {
    const { client } = mockClient(() => ({
      status: 406,
      body: { error: 'protocol_version_unsupported', message: 'unlisted major', details: { protocolVersions: ['1.12'] } },
    }));
    await expect(client.runs.get('t/r1')).rejects.toBeInstanceOf(WopError);
    try {
      await client.runs.get('t/r1');
    } catch (err) {
      const e = err as WopError;
      expect(e.status).toBe(406);
      expect(e.envelope?.error).toBe('protocol_version_unsupported');
      expect(isErrorCode(e.envelope?.error)).toBe(true);
    }
  });
});

describe('header renames (headers.md)', () => {
  it('OpenWOP-Dedup replaces X-Dedup; Idempotency-Key keeps its standard name', async () => {
    const { client, captured } = mockClient(() => ({ status: 201, body: { runId: 't/r', status: 'pending', eventsUrl: '/runs/t%2Fr/events' } }));
    await client.runs.create({ workflowId: 'wf' }, { idempotencyKey: 'idem-1', dedup: 'enforce' });
    const h = captured[0]!.headers;
    expect(h.get('OpenWOP-Dedup')).toBe('enforce');
    expect(h.get('Idempotency-Key')).toBe('idem-1');
    expect(h.has('X-Dedup')).toBe(false);
    expect(captured[0]!.url).toBe('https://host.example/runs');
  });
});

describe('paths are unversioned keys on the bare origin (versioning.md §1.2)', () => {
  it('a sample across every namespace', async () => {
    const { client, captured } = mockClient(() => ({ status: 200, body: {} }));
    await client.discovery.openapi();
    await client.workflows.get('wf');
    await client.runs.get('t/r1');
    await client.runs.compensation('t/r1');
    await client.runs.effects('t/r1');
    await client.agents.getOrgChart();
    await client.tools.list();
    await client.prompts.render({ ref: 'x', variables: {} } as never);
    await client.audit.verify(0, 1);
    await client.content.getSettings();
    await client.webhooks.unregister('wh1');
    await client.host.effectSeams();
    expect(captured.map((c) => new URL(c.url).pathname)).toEqual([
      '/openapi.json',
      '/workflows/wf',
      '/runs/t%2Fr1',
      '/runs/t%2Fr1/compensation',
      '/runs/t%2Fr1/effects',
      '/agents/org-chart',
      '/tools',
      '/prompts:render',
      '/audit/verify',
      '/content/settings',
      '/webhooks/wh1',
      '/host/effect-seams',
    ]);
  });
});

describe('poll cursor (events.md §Poll)', () => {
  it('sends afterSequence, never lastSequence, and reads the closed response', async () => {
    const page = { runId: 't/r1', events: [], lastSequence: -1, status: 'running', isTerminal: false };
    const { client, captured } = mockClient(() => ({ status: 200, body: page }));
    const res = await client.runs.pollEvents('t/r1', { afterSequence: 7, timeoutSeconds: 5 });
    const url = new URL(captured[0]!.url);
    expect(url.pathname).toBe('/runs/t%2Fr1/events/poll');
    expect(url.searchParams.get('afterSequence')).toBe('7');
    expect(url.searchParams.get('timeout')).toBe('5');
    expect(url.searchParams.has('lastSequence')).toBe(false);
    expect(res.lastSequence).toBe(-1);
    expect(res.isTerminal).toBe(false);
  });
});

describe('SSE channels carry the version header and unversioned paths', () => {
  const sse = (frames: string) => {
    const seen: Captured[] = [];
    const f: typeof fetch = async (input, init) => {
      const url = typeof input === 'string' ? input : input instanceof URL ? input.toString() : input.url;
      seen.push({ url, method: init?.method ?? 'GET', headers: new Headers(init?.headers ?? {}) });
      return new Response(frames, { status: 200, headers: { 'content-type': 'text/event-stream' } });
    };
    return { f, seen };
  };

  it('runEvents: /runs/{runId}/events + OpenWOP-Version + Last-Event-ID; batch frames flatten', async () => {
    const { f, seen } = sse('id: 1\nevent: batch\ndata: [{"eventId":"e1","runId":"t/r1","type":"run.started","payload":{},"timestamp":"t","sequence":0,"schemaVersion":1},{"eventId":"e2","runId":"t/r1","type":"run.completed","payload":{},"timestamp":"t","sequence":1,"schemaVersion":1}]\n\n');
    const out = [];
    for await (const ev of streamEvents({ baseUrl: 'https://host.example', apiKey: 'k', protocolVersion: '2.0', fetch: f }, 't/r1', { lastEventId: '0', streamMode: ['updates', 'messages'] })) out.push(ev.type);
    expect(out).toEqual(['run.started', 'run.completed']);
    const url = new URL(seen[0]!.url);
    expect(url.pathname).toBe('/runs/t%2Fr1/events');
    expect(url.searchParams.get('streamMode')).toBe('updates,messages');
    expect(seen[0]!.headers.get('OpenWOP-Version')).toBe('2.0');
    expect(seen[0]!.headers.get('Last-Event-ID')).toBe('0');
  });

  it('hostEvents: /host/events by default, overridable by heartbeat.deliveryChannel', async () => {
    const { f, seen } = sse(': keep-alive\n\nevent: heartbeat.evaluated\ndata: {"type":"heartbeat.evaluated","payload":{"heartbeatId":"h1","status":"ok","changed":false}}\n\n');
    const out = [];
    for await (const ev of streamHostEvents({ baseUrl: 'https://host.example', apiKey: 'k', protocolVersion: '2.0', fetch: f })) out.push(ev.type);
    expect(out).toEqual(['heartbeat.evaluated']);
    expect(new URL(seen[0]!.url).pathname).toBe('/host/events');
    expect(seen[0]!.headers.get('OpenWOP-Version')).toBe('2.0');
    for await (const _ of streamHostEvents({ baseUrl: 'https://host.example', apiKey: 'k', protocolVersion: '2.0', fetch: f }, { path: '/ops/heartbeats' })) void _;
    expect(new URL(seen[1]!.url).pathname).toBe('/ops/heartbeats');
  });
});

describe('generated registries match the vendored corpus', () => {
  it('ERROR_CODES is exactly spec/v2/errors.json, sorted, with the registered status + retriable rows', () => {
    const registry = JSON.parse(readFileSync(resolve(REPO, 'spec/v2/errors.json'), 'utf8')) as { rows: { code: string; httpStatus: number; retriable: boolean }[] };
    const codes = registry.rows.map((r) => r.code).sort();
    expect([...ERROR_CODES]).toEqual(codes);
    expect(ERROR_CODES).toHaveLength(94);
    // rc.1 grew the registry by two rows; the union tracks it, not a hand-kept list.
    expect(isErrorCode('fork_point_invalid')).toBe(true);
    expect(isErrorCode('webhook_url_rejected')).toBe(true);
    expect(ERROR_CODE_HTTP_STATUS['fork_point_invalid']).toBe(422);
    expect(ERROR_CODE_HTTP_STATUS['webhook_url_rejected']).toBe(400);
    expect(HTTP_ERROR_CODES).toBe(ERROR_CODES);
    for (const r of registry.rows) expect(ERROR_CODE_HTTP_STATUS[r.code as (typeof ERROR_CODES)[number]]).toBe(r.httpStatus);
    expect([...RETRIABLE_ERROR_CODES].sort()).toEqual(registry.rows.filter((r) => r.retriable).map((r) => r.code).sort());
    expect(isErrorCode('rate_limited')).toBe(true);
    expect(isRetriableErrorCode('rate_limited')).toBe(true);
    expect(isRetriableErrorCode('not_found')).toBe(false);
    expect(isErrorCode('acme.quota_exceeded')).toBe(false);
    expect(isVendorErrorCode('acme.quota_exceeded')).toBe(true);
    expect(isVendorErrorCode('openwop.reserved')).toBe(false);
  });

  it('capability keys partition the closed root of capabilities.schema.json', () => {
    const schema = JSON.parse(readFileSync(resolve(REPO, 'schemas/v2/capabilities.schema.json'), 'utf8')) as {
      additionalProperties: boolean; required: string[]; properties: Record<string, { required?: string[] }>;
    };
    expect(schema.additionalProperties).toBe(false);
    expect(schema.required.sort()).toEqual(['preferredVersion', 'protocolVersions']);
    const all = Object.keys(schema.properties).sort();
    expect([...CAPABILITY_FAMILY_KEYS, ...CAPABILITY_METADATA_KEYS].sort()).toEqual(all);
    for (const k of CAPABILITY_FAMILY_KEYS) expect(schema.properties[k]?.required).toEqual(expect.arrayContaining(['status', 'since', 'witness']));
  });
});
